auth.go

```diff
@@ -0,0 +1,218 @@
+}
+
+func NewAuthManager(store Store, keyname string, idsrc IdentityDataGetter) AuthManager {
+	ks := NewKeystore(store)
+	if ks == nil {
+		return nil
+	}
+
+	mgr := &MenderAuthManager{
+		store:    store,
+		keystore: ks,
+		keyname:  keyname,
+		idsrc:    idsrc,
+		seqnum:   NewFileSeqnum(authSeqName, store),
+	}
+
+	if err := ks.Load(keyname); err != nil && IsNoKeys(err) == false {
+		log.Errorf("failed to load device keys: %s", err)
+		return nil
+	}
+
+	return mgr
+}
+
+func (m *MenderAuthManager) IsValid() bool {
+	adata, err := m.AuthCode()
+	if err != nil {
+		return false
+	}
+
+	if adata == noAuthCode {
+		return false
+	}
+
+	// TODO check if JWT is valid?
+
+	return true
+}
+
+func (m *MenderAuthManager) MakeAuthRequest() (*AuthRequest, error) {
+
+	var err error
+	authd := AuthReqData{}
+
+	idata, err := m.idsrc.Get()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to obtain identity data")
+	}
+
+	authd.IdData = idata
+
+	// fill device public key
+	authd.Pubkey = m.keystore.PublicPEM()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to obtain device public key")
+	}
+
+	tentok, err := m.store.ReadAll(authTenantTokenName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read tenant token")
+	}
+	log.Debugf("tenant token: %s", tentok)
+
+	// fill tenant token
+	authd.TenantToken = string(tentok)
+
+	// fetch sequence number
+	num, err := m.seqnum.Get()
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to obtain sequence number")
+	}
+	authd.SeqNumber = num
+
+	log.Debugf("authorization data: %v", authd)
+
+	databuf := &bytes.Buffer{}
+	enc := json.NewEncoder(databuf)
+
+	err = enc.Encode(&authd)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to encode auth request")
+	}
+
+	reqdata := databuf.Bytes()
+
+	// generate signature
+	sig, err := m.keystore.Sign(reqdata)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to sign auth request")
+	}
+
+	return &AuthRequest{
+		Data:      reqdata,
+		Code:      AuthCode(tentok),
+		Signature: sig,
+	}, nil
+}
+
+func (m *MenderAuthManager) RecvAuthResponse(data []byte) error {
+	if len(data) == 0 {
+		return errors.New("empty auth response data")
+	}
+
+	if err := m.store.WriteAll(authTokenName, data); err != nil {
+		return errors.Wrapf(err, "failed to save auth token")
+	}
+	return nil
+}
+
+func (m *MenderAuthManager) AuthCode() (AuthCode, error) {
+	data, err := m.store.ReadAll(authTokenName)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return noAuthCode, nil
+		}
+		return noAuthCode, errors.Wrapf(err, "failed to read auth token data")
+	}
+
+	return AuthCode(data), nil
+}
+
+func (m *MenderAuthManager) HasKey() bool {
+	if m.keystore.Private() != nil {
+		return true
+	}
+	return false
+}
+
+func (m *MenderAuthManager) GenerateKey() error {
+	if err := m.keystore.Generate(); err != nil {
+		log.Errorf("failed to generate device key: %v", err)
+		return errors.Wrapf(err, "failed to generate device key")
```

Comment:
I think we need to agree on some convention here. In more places we are rather just returning error without logging anything. 
From the other hand maybe there is nothing wrong with logging the error twice.